kernel/proc.c

```diff
@@ -457,28 +510,31 @@ scheduler(void)
     intr_on();
     intr_off();
 
-    int found = 0;
-    for(p = proc; p < &proc[NPROC]; p++) {
-      acquire(&p->lock);
-      if(p->state == RUNNABLE) {
-        // Switch to chosen process.  It is the process's job
-        // to release its lock and then reacquire it
-        // before jumping back to us.
-        p->state = RUNNING;
-        c->proc = p;
-        swtch(&c->context, &p->context);
-
-        // Process is done running for now.
-        // It should have changed its p->state before coming back.
-        c->proc = 0;
-        found = 1;
-      }
-      release(&p->lock);
-    }
-    if(found == 0) {
+
+    p = get_runnable_min_pass_proc_locked();
+
+    if(p == 0) {
       // nothing to run; stop running on this core until an interrupt.
       asm volatile("wfi");
+      continue;
     }
+
+    // Switch to chosen process.  It is the process's job
+    // to release its lock and then reacquire it
+    // before jumping back to us.
+    p->state = RUNNING;
+    c->proc = p;
+    
+    swtch(&c->context, &p->context);
+
+    p->pass += MAX_TICKETS / p->tickets;
+    cached_min_pass = p->pass;
+    p->ticks++;
```

Comment:
현재 구현에선 같은 min_pass 값을 가진 프로세스가 여러개 있을 수 있어서, 경우에 따라 특정 프로세스가 cached_min_pass보다 낮은 값의 min_pass를 가질수도 있게됩니다. 아래와 같이 min_pass 를 찾은 즉시 cached_min_pass 를 업데이트하도록 하면, cached_min_pass보다 낮은 pass 값을 가지는 프로세스가 없음이 보장됩니다.

```suggestion
    cached_min_pass = p->pass;

    // Switch to chosen process.  It is the process's job
    // to release its lock and then reacquire it
    // before jumping back to us.
    p->state = RUNNING;
    c->proc = p;
    
    swtch(&c->context, &p->context);

    p->pass += MAX_TICKETS / p->tickets;
    p->ticks++;
```

다만 이렇게하면 cached_min_pass가 실제 min_pass와 같지 않고 더 낮은 경우가 있을 수 있습니다. 지금 프로세스 목록이 배열로 관리되고있고, pass로 정렬해놓은 힙이나 트리가 어디에 따로 없어서 min_pass 계산이 O(n)으로 비싸다보니 이런 문제가 생기는건데,

1. cached_min_pass가 순간적으로 실제 min_pass보다 높아지는것을 허용하기 (기존 구현)
2. cached_min_pass가 순간적으로 실제 min_pass보다 낮아지는것을 허용하기 (제 제안)

둘중 하나를 골라야하는데 저는 1번보다는 2번이 나은것같습니다.

Reply:
아 생각못했던 부분이네요. 저도 새 프로세스 pass 값이 최소값으로 보장되는 게 더 중요하다 생각해서 제안해주신대로 수정하겠습니다!